redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/FqType.kt

```diff
@@ -0,0 +1,159 @@
+import kotlin.reflect.KTypeProjection
+import kotlin.reflect.KVariance.IN
+import kotlin.reflect.KVariance.INVARIANT
+import kotlin.reflect.KVariance.OUT
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+public data class FqType(
+  /**
+   * The package name followed by one or more simple names. The package may be empty.
+   * A star projection is represented by an empty package and a single "*" simple name.
+   */
+  val names: List<String>,
+  val variance: Variance = Invariant,
+  val parameterTypes: List<FqType> = emptyList(),
+  val nullable: Boolean = false,
+) {
+  init {
+    require(names.size >= 2) {
+      "At least two names are required: package and a simple name: $names"
+    }
+    if (names[1] == "*") {
+      require(names[0] == "") {
+        "Star projection must use empty package name: ${names[0]}"
+      }
+      require(names.size == 2) {
+        "Star projection cannot have nested types: ${names.drop(1)}"
+      }
+      require(variance == Invariant) {
+        "Star projection must be Invariant: $variance"
+      }
+      require(parameterTypes.isEmpty()) {
+        "Star projection must not have parameter types: $parameterTypes"
+      }
+      require(!nullable) {
+        "Star projection must not be nullable"
+      }
+    }
+  }
+
+  public enum class Variance {
+    Invariant,
+
+    @SerialName("in")
+    In,
+
+    @SerialName("out")
+    Out,
+  }
+
+  override fun toString(): String = buildString {
+    when (variance) {
+      Invariant -> Unit // Nothing to do.
+      In -> append("in ")
+      Out -> append("out ")
+    }
+    if (names[0] != "") {
+      append(names[0])
+      append('.')
+    }
+    append(names[1])
+    for (i in 2 until names.size) {
+      append('.')
+      append(names[i])
+    }
+    if (parameterTypes.isNotEmpty()) {
+      append('<')
+      parameterTypes.joinTo(this, separator = ", ")
+      append('>')
+    }
+    if (nullable) {
+      append('?')
+    }
+  }
+
+  public companion object {
+    public val Star: FqType = FqType(listOf("", "*"))
+  }
+}
+
+internal fun KClass<*>.toFqType(): FqType {
+  var qualifiedName = qualifiedName ?: throw AssertionError(this)
+  val names = when (qualifiedName) {
+    "kotlin.Boolean.Companion" -> listOf("kotlin", "Boolean", "Companion")
+    "kotlin.Byte.Companion" -> listOf("kotlin", "Byte", "Companion")
+    "kotlin.Char.Companion" -> listOf("kotlin", "Char", "Companion")
+    "kotlin.Double.Companion" -> listOf("kotlin", "Double", "Companion")
+    "kotlin.Enum.Companion" -> listOf("kotlin", "Enum", "Companion")
+    "kotlin.Float.Companion" -> listOf("kotlin", "Float", "Companion")
+    "kotlin.Int.Companion" -> listOf("kotlin", "Int", "Companion")
+    "kotlin.Long.Companion" -> listOf("kotlin", "Long", "Companion")
+    "kotlin.Short.Companion" -> listOf("kotlin", "Short", "Companion")
+    "kotlin.String.Companion" -> listOf("kotlin", "String", "Companion")
```

Comment:
These are a list of Kotlin types whose enclosing class name is a type that is mapped to a JVM class. Thus, the class backing the nested Kotlin type does not have an enclosing class (i.e., a parent) and the below algorithm will fail.

In other words, `kotlin.Boolean.Companion`'s `.java` class is `kotlin.internal.jvm.BooleanCompanionObject` or something, and it's `enclosingClass` is null.

Reply:
Put as a comment in the source.